### ISSUE TITLE
Checkr Background Check Workflow Does Not Check For Email Address Before Submitting

### ISSUE THREAD

**Issue by JimMichael** [SparkDevNetwork/Rock#4932]
### Description

The Checkr Background Check workflow that is launched from the Actions menu on a person does not verify that the person (candidate) has any email address on their profile before submitting the request. This guarantees such requests will fail, with no obvious way for the requester to know that's what happened or how to fix/re-submit it.

### Steps to Reproduce

1. Go to a person record with NO email address on file
2. Click on Actions | Background Check (this assumes the system is configured for Checkr background checks, not PMM)
3. Submit the request
4. Observe that the workflow instantly errors-out, because there is no email address for this candidate and Checkr cannot function (contact the candidate) without such address.

**Expected behavior:**

Because an email address on a candidate is REQUIRED for Checkr to function, the workflow should be warning the requester that no address exists for the candidate, and prevent the submission of the workflow until such an address exists (with an explanatory error)

**Actual behavior:**

The Checkr workflow allows the submission (request) for candidates that have no address, which instantly fail and give the requester no obvious steps for resolution. Instead, Background Check Admins get a link on their homepage that says "Background Check: [name] (Request Error)" but this link takes you to a form in the workflow where you can only re-submit it... it doesn't tell you that the error is due to a missing email address on the profile, and even if it did you still can't edit it here. The requester gets no notification at all.

It would be much easier/simpler/better to just have the workflow refuse to submit if there is no email address for the candidate, with a good error/explanation telling the requester what the problem is.

While we realize this is a pretty trivial thing to add to the Checkr workflow ourselves, we are hesitant to edit such a Core workflow for fear our modifications could break future changes to the workflow from core.

### Versions

* **Rock Version:** All versions that have Checkr workflow

**Comment by crayzd92** [comment]
@JimMichael is this essentially the same as https://community.rockrms.com/ideas/1675? If so, we are working on getting that into 13.2 for Beta.

**Comment by JimMichael** [comment]
Yes, sorry! That idea didn't seem to be getting any traction and after this happened again multiple times recently I decided to put an issue in for it as the behavior is broken by letting you submit the form (request) without all necessary data being in place.

I guess this can be closed, then!

**Comment by crayzd92** [comment]
@JimMichael The idea has been updated and this is in v13.2

**Comment by JimMichael** [comment]
Perhaps I'm missing something (I hope?) but I'm struggling to see how this change in 13.2 is considered a "fix" for the issue I brought up with the Checkr background check when the candidate has no email address. 

As far as I can tell, the only change was to add a _Set Person Email Valid Attribute_ action, which determines if the Person has an email address (good) and sets a flad, and then a subsequent _Person Email Valid_ action that simply **completes the workflow** if the flag is true (seriously?)

![image](https://user-images.githubusercontent.com/3903408/159199834-0c78d6cc-fd35-4348-8a58-428b7be1b028.png)

The result is that a person submitting a Checker Background check for a Person receives this suuuppper helpful error message
![image](https://user-images.githubusercontent.com/3903408/159199904-517214a1-2cb4-4e91-8bab-2dedae6bb5ea.png)

that tells them absolutely nothing useful (like, "the person you are trying to submit a BGC for does not have an email address" or something similar. I never like to jump to conclusions, but again... I'm struggling to understand how this fix passed the "we implemented your useful suggestion" test. It's almost funny. Almost.

**Comment by JimMichael** [comment]
Also just wanted to clarify that the above came off harsher than I'd intended, so I apologize if it came off that way. I should know better than to post at night when I'm tired :-(  I just wanted to find out why this is considered fixing the core issue of the enhancement, when (again, unless I'm missing something... which is ENTIRELY possible), it does not.

**Comment by crayzd92** [comment]
@JimMichael no worries. I am looking into your concerns with the development team here and will get back to you here. In the meantime here is the [code change](https://github.com/SparkDevNetwork/Rock/blob/develop/Rock.Migrations/Migrations/Version%2013.0/Version%201.13.2/202203082129530_BackgroundCheckWorkflowEmailActions.cs) for this request. Since it was in a migration file it didn't end up as a commit link here. We'll do better.

**Comment by dynamiclynk** [comment]
@JimMichael @crayzd92  Okay I have stood a new 13.2 instance and it works as expected. I am researching to see why it is busted on in your installation. Is this a new instance of 13.2 or an upgrade in your case? 

As expected:
![image](https://user-images.githubusercontent.com/621804/159519849-22f632e8-43c3-41cc-a9f0-89e4c997400b.png)

Actual flow:
![image](https://user-images.githubusercontent.com/621804/159520042-a63ebf6c-2a2b-424a-8dfc-5c9318d1c6d1.png)

Result:
![image](https://user-images.githubusercontent.com/621804/159520259-8c75226a-fcbb-4724-baa4-de1093c91c39.png)

**Comment by JimMichael** [comment]
Thanks for this info, @dynamiclynk ! So glad to know that it's broken somehow and that what I'm seeing wasn't the intended fix. 

Just in case it matters or helps... after looking at our Checkr workflow I can see it's been modified slightly from stock... we added an attribute to act as a flag, and an additional action so we can launch the workflow from the Actions menu *or* a Connection Request. 

Maybe that somehow broke the migration? But that wouldn't explain why it's also broken on prealpha (which is where I got my screen shots above from, for what it's worth).

**Comment by ethan-sparkdevnetwork** [comment]
@JimMichael I found an issue where the workflow didn't have the PersonEmailValid attribute set when checking the email. The result was that every request would say the email address was missing or invalid.

If you're still seeing incorrect functionality after 13.7 can you send me a couple of screenshots of the workflow in Edit mode:
1.  The folded list of all of the actions in the Initial Request Activity
2. The Expanded view of the Set PersonEmailValid Attribute Action

If you want to test out a script I can send that to you as well.

**Comment by JimMichael** [comment]
Thanks for fixing this! We've known about it quite a while in chat and have walked several users seeing the problem through updating that action manually, but having it fixed automatically is much better ;-)
